Airspeed: preflight check for bad offset, fixed calls to preflight ch…

AndreasAntener opened the pull request:
…ecks (vtol & airspeed)

This is best used with a pitot tube cover. Without a cover and lots of wind on the ground the check will report a calibration issue. Although it's annoying to do more airspeed calibrations, this effectively got rid of all offset issues for us.

Fixes #5059 (I think)
Fixes #5631

@dagar @julianoes please have a look

dagar commented:
I'm not sure I understand how this is different. Isn't checking the differential pressure effectively the same as the current indicated airspeed check? Except ~5 m/s instead of 6

AndreasAntener commented:
Differences are:
- checks also negative offsets (and yes I made the margin smaller)
- actually fails preflight check now
- ignores check when arming (tube removed, could read airspeed at that time)
- fixes calls to check which were wrong before (for example it didn't check airspeed for VTOLs)

dagar commented:
Ah right it can be negative. Thanks for the explanation, I'll do some testing.

dagar commented:
Airspeed calibration fails if you blow too hard.

AndreasAntener commented:
Huh, do you have an idea why? I'll give that a try too

julianoes commented on src/modules/commander/PreflightCheck.cpp:
```diff
@@ -338,11 +338,24 @@ static bool airspeedCheck(orb_advert_t *mavlink_log_pub, bool optional, bool rep
+	if (fabsf(airspeed.differential_pressure_filtered_pa) > 15.0f && !prearm) {
 		if (report_fail) {
-			mavlink_log_critical(mavlink_log_pub, "AIRSPEED WARNING: WIND OR CALIBRATION ISSUE");
+			mavlink_log_critical(mavlink_log_pub, "PREFLIGHT FAIL: AIRSPEED CALIBRATION ISSUE");
```
I would say what's wrong instead of "AIRSPEED CALIBRATION ISSUE"

AndreasAntener replied:
Hm, my idea was to point to the thing that will resolve the issue. Because something like "large airspeed offset" doesn't really help I think. Other suggestions?

julianoes commented on src/modules/commander/commander_tests/state_machine_helper_test.cpp:
```diff
@@ -293,10 +293,10 @@ bool StateMachineHelperTest::armingStateTransitionTest(void)
-        		5.0f, check_gps);
+				false /* no pre-arm checks */,
+				nullptr /* no mavlink_log_pub */,
+				&status_flags,
```
Indenting looks crooked, at least on github.

AndreasAntener replied:
It was wrong before, the git white space check complained to me :o

julianoes commented on src/modules/commander/commander_tests/state_machine_helper_test.cpp:
```diff
@@ -293,10 +293,10 @@ bool StateMachineHelperTest::armingStateTransitionTest(void)
+				false /* no pre-arm checks */,
+				nullptr /* no mavlink_log_pub */,
+				&status_flags,
+				5.0f, check_gps, 2e6);
```
What are these two values? Can you add a comment?

AndreasAntener commented:
@julianoes addressed your comments
@dagar calibration succeeds for me even with max pressure (I use a tube cover that I can compress)

I fixed this up (I missed a commit for VTOL). Also @julianoes , please have a look at the last commit I added. You didn't get the failures reported via MAVLink because they got flagged as reported when they were actually not. Reproducible when telemetry or MAVLink via USB is running before the hotplug timeout is over.

dagar commented:
Tried again and blowing too hard (> 5 m/s) doesn't fail calibration, but you get the critical preflight fail message which is confusing. Must be the filtered value immediately after calibration finishes?

What if the end of airspeed calibration (after the beep for successful positive pressure) we wait for the filtered pressure to drop below the preflight threshold in a reasonable time and fail calibration if it doesn't?

AndreasAntener commented:
Oh I see. I would probably just delay the end of the calibration a fraction, to not duplicate the preflight check. Although you could argue that this check belongs into the calibration anyway.

AndreasAntener commented:
@dagar I was able to reproduce the issue and added a commit that should prevent this in 99% of the cases. Unless the operator creates airflow for a long time it should be fine.

dagar commented:
Thanks!

On Fri, Dec 2, 2016 at 5:55 AM, Andreas Daniel Antener <
notifications@github.com> wrote:

> @dagar <https://github.com/dagar> I was able to reproduce the issue and
> added a commit that should prevent this in 99% of the cases. Unless the
> operator creates airflow for a long time it should be fine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/PX4/Firmware/pull/5851#issuecomment-264428252>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAFK6DWqAPGLBOE43-1tbWPz37pJUNCBks5rD_kqgaJpZM4Kw10C>
> .
>
